### PR TITLE
feat: add mcp support for cluster planes

### PR DESCRIPTION
**Pull request by yashodgayashan**
## Purpose

Add MCP (Model Context Protocol) tool support for cluster-scoped plane resources — ClusterDataPlane, ClusterBuildPlane, and ClusterObservabilityPlane. These resources already have REST API endpoints and services but were not accessible via MCP, meaning MCP clients (e.g., AI assistants, IDE integrations) couldn't manage or inspect shared platform infrastructure.  

## Approach

Mirrored the existing namespace-scoped MCP tool pattern (e.g., list_dataplanes, get_dataplane) but without the namespace_name parameter since cluster-scoped resources aren't namespaced.                                                    
                                                                                  
**5 new MCP tools added to the InfrastructureToolset:**                             

1. Tool: list_cluster_dataplanes                                                   
    Description: List all cluster-scoped data planes                                                                  
2. Tool: get_cluster_dataplane                                                     
  Description: Get details of a specific cluster data plane by name
3. Tool: create_cluster_dataplane
  Description: Create a new cluster data plane (with full gateway/port/observability config)
4. Tool: list_cluster_buildplanes
  Description: List all cluster-scoped build planes
5. Tool: list_cluster_observability_planes
  Description: List all cluster-scoped observability planes

## Checklist
- [ ] Tests added or updated (unit, integration, etc.)
- [ ] Samples updated (if applicable)

## Remarks
- The create_cluster_dataplane tool hardcodes observabilityPlaneRef.kind to ClusterObservabilityPlane since cluster-scoped planes can only reference cluster-scoped observability planes (enforced by CEL validation on the CRD).
- No changes to services, handlers, or wiring — the MCP handlers delegate to existing service methods (ClusterDataPlaneService, ClusterBuildPlaneService, ClusterObservabilityPlaneService) that were already wired into the MCPHandler.
- Port fields (e.g., public_http_port) use inline map[string]any{"type": "integer", ...} since no intProperty helper exists in the tools package.

## PR Summary: Add MCP Support for Cluster Planes

Scope: Add MCP (Model Context Protocol) tool support for cluster-scoped plane resources so MCP clients can manage/inspect cluster-scoped infrastructure (ClusterDataPlane, ClusterBuildPlane, ClusterObservabilityPlane). MCP handlers are added and tools registered; handlers delegate to existing Cluster*Service implementations.

Mini breakdown — changed files by top-level folder
- internal/: 3 files
  - internal/openchoreo-api/mcphandlers/clusterbuildplanes.go (+25)
  - internal/openchoreo-api/mcphandlers/clusterdataplanes.go (+41)
  - internal/openchoreo-api/mcphandlers/clusterobservabilityplanes.go (+25)
- pkg/: 5 files
  - pkg/mcp/tools/infrastructure.go (+173)
  - pkg/mcp/tools/infrastructure_specs_test.go (+73)
  - pkg/mcp/tools/mock_test.go (+27)
  - pkg/mcp/tools/register.go (+5)
  - pkg/mcp/tools/types.go (+18)

Total: 8 files changed, +387 lines added.

API / CRD surface changes (explicit)
- New MCP tools (cluster-scoped) added (additive):
  1. list_cluster_dataplanes — ListClusterDataPlanes (no args)
  2. get_cluster_dataplane — GetClusterDataPlane (requires cdp_name)
  3. create_cluster_dataplane — CreateClusterDataPlane (create ClusterDataPlane). Input surface includes: name, plane_id, cluster_agent_client_ca, public_virtual_host, organization_virtual_host (required); optional display_name, description, public_http_port, public_https_port, organization_http_port, organization_https_port, observability_plane_ref. Implementation forcibly sets observabilityPlaneRef.kind = "ClusterObservabilityPlane".
  4. list_cluster_buildplanes — ListClusterBuildPlanes (no args)
  5. list_cluster_observability_planes — ListClusterObservabilityPlanes (no args)
- New interface: ClusterPlaneHandler with five methods; infrastructure tool registration updated to register these tools only when implemented.
Compatibility risk: LOW — purely additive MCP tool registrations and interface additions; no changes to CRD schemas, REST APIs, or services. Handlers delegate to existing services.

Tests added / updated
- pkg/mcp/tools/infrastructure_specs_test.go — added 5 new tool specs for the cluster-scoped tools (+73 lines). Note: the new specs are duplicated twice in the file (potential accidental duplication).
- pkg/mcp/tools/mock_test.go — added mock implementations for the 5 new methods (+27 lines).

Tests still missing / critical paths lacking coverage
- No unit tests added for new MCP handler implementations under internal/openchoreo-api/mcphandlers (List/Get/Create handlers).
- No unit tests for CreateClusterDataPlane validation logic (observability_plane_ref.kind enforcement, port typing).
- No integration/e2e tests for end-to-end MCP tool invocation or service interaction.
- Samples and documentation not updated; checklist items remain unchecked.

Risk hotspots (short reasons)
1. Handler unit tests missing — High: MCP handler behavior (response shape, error wrapping) is untested and could regress MCP clients.
2. Hardcoded observabilityPlaneRef.kind — Medium: relies on CRD CEL validation for cluster-scoped refs; missing tests may hide edge cases or user errors.
3. Port schema/type handling — Medium: ports modeled via inline map[string]any with integer-type schema (no intProperty helper); risks type-safety or client parsing issues.
4. AuthN/AuthZ assumptions — Low/Medium: handlers call existing services (assumed to enforce authz); MCP-layer authorization behavior not validated here.
5. Test duplication in specs — Low/Medium: duplicated specs increase maintenance burden and may mask true test coverage.

Notes / recommendations
- Add unit tests for mcphandlers (List/Get/Create) and for CreateClusterDataPlane validation (observability ref and ports).
- Deduplicate specs in infrastructure_specs_test.go.
- Add integration/e2e tests and update samples/docs to validate MCP tool end-to-end behavior.

### CODE REVIEW

File: pkg/mcp/tools/infrastructure.go [hunk 1]
```diff
@@ -354,3 +354,145 @@ func (t *Toolsets) RegisterGetComponentWorkflowSchemaOrgLevel(s *mcp.Server) {
 		return handleToolResult(result, err)
 	})
 }
+
+func (t *Toolsets) RegisterListClusterDataPlanes(s *mcp.Server) {
+	mcp.AddTool(s, &mcp.Tool{
+		Name: "list_cluster_dataplanes",
```
Comment (chathuranga95) [inline pkg/mcp/tools/infrastructure.go:369]: Tried invoking this tools and I get the following error.
```
failed to list cluster dataplanes: clusterdataplanes.openchoreo.dev is forbidden: User \"system:serviceaccount:openchoreo-control-plane:openchoreo-api\" cannot list resource \"clusterdataplanes\" in API group \"openchoreo.dev\" at the cluster scope
```

Any special permission needed to invoke the tools?

Comment (yashodgayashan) [reply pkg/mcp/tools/infrastructure.go:369]: Good Catch. Fixed with https://github.com/openchoreo/openchoreo/pull/1945

File: pkg/mcp/tools/infrastructure.go [hunk 2]
```diff
@@ -354,3 +354,145 @@ func (t *Toolsets) RegisterGetComponentWorkflowSchemaOrgLevel(s *mcp.Server) {
...
+		InputSchema: createSchema(map[string]any{
+			"cdp_name": stringProperty("Use list_cluster_dataplanes to discover valid names"),
+		}, []string{"cdp_name"}),
+	}, func(ctx context.Context, req *mcp.CallToolRequest, args struct {
+		CdpName string `json:"cdp_name"`
+	}) (*mcp.CallToolResult, any, error) {
+		result, err := t.InfrastructureToolset.GetClusterDataPlane(ctx, args.CdpName)
+		return handleToolResult(result, err)
+	})
+}
```
[the remark this answers is not in the document]
Comment (yashodgayashan) [reply pkg/mcp/tools/infrastructure.go:403]: This has been fixed

**Review by nilushancosta — APPROVED** [review]
